python/ray/actor.py

```diff
@@ -204,8 +204,55 @@ class ActorClass(object):
             each actor method.
     """
 
-    def __init__(self, modified_class, class_id, max_reconstructions, num_cpus,
-                 num_gpus, memory, object_store_memory, resources):
+    def __init__(cls, name, bases, attr):
+        """Prevents users from directly inheriting from an ActorClass.
+
+        This will be called when a class is defined with an ActorClass object
+        as one of its base classes. To intentionally construct an ActorClass,
+        use the 'from_modified_class' classmethod.
+
+        Raises:
+            TypeError: Always.
+        """
+        for base in bases:
+            if isinstance(base, ActorClass):
+                raise TypeError("Attempted to define subclass '{}' of actor "
+                                "class '{}'. Inheriting from actor classes is "
+                                "not currently supported. You can instead "
+                                "inherit from a non-actor base class and make "
+                                "the derived class an actor class (with "
+                                "@ray.remote).".format(name, base._class_name))
+
+        # This shouldn't be reached because one of the base classes must be
+        # an actor class if this was called, but raise a more generic error
+        # just in case.
+        raise TypeError("Attempted to define subclass '{}' of an actor class. "
+                        "Inheriting from actor classes is not currently "
+                        "supported. You can instead inherit from a non-actor "
+                        "base class and make the derived class an actor class "
+                        "(with @ray.remote).".format(name, base._class_name))
+
+    def __call__(self, *args, **kwargs):
+        """Prevents users from directly instantiating an ActorClass.
+
+        This will be called instead of __init__ when 'ActorClass()' is executed
+        because an is an object rather than a metaobject. To properly
+        instantiated a remote actor, use 'ActorClass.remote()'.
+
+        Raises:
+            Exception: Always.
+        """
+        raise Exception("Actors cannot be instantiated directly. "
+                        "Instead of '{}()', use '{}.remote()'.".format(
+                            self._class_name, self._class_name))
+
+    @classmethod
+    def from_modified_class(cls, modified_class, class_id, max_reconstructions,
```

Comment:
This should probably be hidden from the user. Right now, if I define an actor class `Foo`, and type `Foo.` and then hit "tab" in IPython, it will show `from_modified_class` as an option, see the below screenshot. Same goes for `class_id`.
 
<img width="755" alt="Screen Shot 2019-09-02 at 6 03 32 PM" src="https://user-images.githubusercontent.com/249517/64137282-23c9af80-cdac-11e9-8aac-dc747ce85aeb.png">

You can achieve this by changing the name to `_from_modified_class`.

Reply:
Good idea, changed.